pool.go

```diff
@@ -0,0 +1,273 @@
+			p.handleError(fmt.Errorf("connecting to %s: %w", conn.addr, err))
+			p.wg.Add(1)
+			go p.recreateConnection(conn)
+			continue
+		}
+
+		p.connections = append(p.connections, conn)
+	}
+
+	if len(p.connections) < p.Opts.MinConnections {
+		return fmt.Errorf("minimum %d connections is required, established: %d", p.Opts.MinConnections, len(p.connections))
+	}
+
+	return nil
+}
+
+// Connections returns copy of all connections from the pool
+func (p *Pool) Connections() []*Connection {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// create new slice, as p.connections can be changed
+	var conns []*Connection
+	conns = append(conns, p.connections...)
+
+	return conns
+}
+
+// FilterFunc is a function to filter connections
+type FilterFunc func(*Connection) bool
+
+// Get returns filtered connection from the pool
+func (p *Pool) Get() (*Connection, error) {
+	p.mu.Lock()
+	if p.isClosed {
+		p.mu.Unlock()
+		return nil, errors.New("pool is closed")
+	}
+	p.mu.Unlock()
+
+	// filtered connections
+	conns := p.filteredConnections()
+
+	if len(conns) == 0 {
+		return nil, errors.New("no (filtered) connections in the pool")
+	}
+
+	n := atomic.AddUint32(&p.connIndex, 1)
+
+	return conns[(int(n)-1)%len(conns)], nil
+}
+
+// when connection is closed, remove it from the pool of connections and start
+// goroutine to create new connection for the same address
+func (p *Pool) handleClosedConnection(closedConn *Connection) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	var connIndex = -1
+	for i, conn := range p.connections {
+		if conn == closedConn {
+			connIndex = i
+			break
+		}
+	}
+
+	// somehow we didn't find closed connection in the pool
+	if connIndex < 0 {
+		p.handleError(errors.New("closed connection was not found in the pool"))
+		return
+	}
+
+	p.connections[connIndex] = p.connections[len(p.connections)-1] // Copy last element to index connIndex.
+	p.connections[len(p.connections)-1] = nil                      // Erase last element
+	p.connections = p.connections[:len(p.connections)-1]           // Truncate slice.
+
+	// if pool was closed, don't start recreate goroutine
+	// should we return earlier and keep connection in the pool as it will be closed anyway?
+	if p.isClosed {
+		return
+	}
+
+	// initiate goroutine to reconnect to closedConn.Addr
+	p.wg.Add(1)
+	go p.recreateConnection(closedConn)
+}
+
+func (p *Pool) recreateConnection(closedConn *Connection) {
+	defer p.wg.Done()
+
+	var conn *Connection
+	var err error
+	for {
+
+		conn, err = p.Factory(closedConn.addr)
+		if err != nil {
+			p.handleError(fmt.Errorf("failed to re-create connection for %s: %w", closedConn.addr, err))
+			return
+		}
+
+		// set own handler when connection is closed
+		conn.SetOptions(ConnectionClosedHandler(p.handleClosedConnection))
+
+		err = conn.Connect()
+
+		if err == nil {
+			break
+		}
+
+		p.handleError(fmt.Errorf("failed to reconnect to %s: %w", conn.addr, err))
+		select {
+		case <-time.After(p.Opts.ReconnectWait):
+			continue
+		case <-p.Done():
```

Comment:
I think this is ok because all calls to recreateConnection are after calls that do
if p.isClosed {
		return
}

That being said, I think the above lines should be included here as if someone does not check if it is closed before calling  recreateConnection case <-p.Done() would be problematic if it was already Closed

Reply:
if it channel was closed, `<-p.Done()` will immediately return the same value that was returned last time.

Just an example, if we have 10 goroutines, each of them reads from `done`: `<- done`. When in main goroutine we `close(done)`, all goroutines will read the value from the channel.